core/unit_test/tstParticleList.hpp

```diff
@@ -0,0 +1,227 @@
+ * This file is part of the Cabana library. Cabana is distributed under a   *
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <Kokkos_Core.hpp>
+
+#include <Cabana_DeepCopy.hpp>
+#include <Cabana_Fields.hpp>
+#include <Cabana_ParticleList.hpp>
+
+#include <gtest/gtest.h>
+
+namespace Test
+{
+//---------------------------------------------------------------------------//
+// Fields.
+struct Foo : Cabana::Field::Scalar<double>
+{
+    static std::string label() { return "foo"; }
+};
+
+struct Bar : Cabana::Field::Matrix<double, 3, 3>
+{
+    static std::string label() { return "bar"; }
+};
+
+//---------------------------------------------------------------------------//
+void particleListTest()
+{
+    // Make a particle list.
+    using list_type =
+        Cabana::ParticleList<TEST_MEMSPACE, Cabana::Field::Position<3>, Foo,
+                             Cabana::Field::CommRank, Bar>;
+    list_type particles( "test_particles" );
+
+    // Resize the aosoa.
+    auto& aosoa = particles.aosoa();
+    std::size_t num_p = 10;
+    aosoa.resize( num_p );
+    EXPECT_EQ( particles.aosoa().size(), 10 );
+
+    // Populate fields.
+    auto px = particles.slice( Cabana::Field::Position<3>() );
+    auto pm = particles.slice( Foo() );
+    auto pc = particles.slice( Cabana::Field::CommRank() );
+    auto pf = particles.slice( Bar() );
+
+    Cabana::deep_copy( px, 1.23 );
+    Cabana::deep_copy( pm, 3.3 );
+    Cabana::deep_copy( pc, 5 );
+    Cabana::deep_copy( pf, -1.2 );
+
+    // Check the slices.
+    EXPECT_EQ( px.label(), "position" );
+    EXPECT_EQ( pm.label(), "foo" );
+    EXPECT_EQ( pc.label(), "comm_rank" );
+    EXPECT_EQ( pf.label(), "bar" );
+
+    // Check deep copy.
+    auto aosoa_host =
+        Cabana::create_mirror_view_and_copy( Kokkos::HostSpace(), aosoa );
+    auto px_h = Cabana::slice<0>( aosoa_host );
+    auto pm_h = Cabana::slice<1>( aosoa_host );
+    auto pc_h = Cabana::slice<2>( aosoa_host );
+    auto pf_h = Cabana::slice<3>( aosoa_host );
+    for ( std::size_t p = 0; p < num_p; ++p )
+    {
+        for ( int d = 0; d < 3; ++d )
+            EXPECT_DOUBLE_EQ( px_h( p, d ), 1.23 );
+
+        EXPECT_DOUBLE_EQ( pm_h( p ), 3.3 );
+
+        EXPECT_EQ( pc_h( p ), 5 );
+
+        for ( int i = 0; i < 3; ++i )
+            for ( int j = 0; j < 3; ++j )
+                EXPECT_DOUBLE_EQ( pf_h( p, i, j ), -1.2 );
+    }
+
+    // Locally modify.
+    Kokkos::parallel_for(
+        "modify", Kokkos::RangePolicy<TEST_EXECSPACE>( 0, num_p ),
+        KOKKOS_LAMBDA( const int p ) {
+            typename list_type::particle_type particle( aosoa.getTuple( p ) );
```

Comment:
Yeah I'd like some helper functions to do this on device. We should be able to pass the particle list to the device and get the AoSoA out of it there.

Reply:
I added these, but maybe `get`/`set` is confusing with the existing free functions